oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticSuggestIterator.java

```diff
@@ -0,0 +1,116 @@
+ * a completion result.
+ */
+class ElasticSuggestIterator implements Iterator<FulltextResultRow> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ElasticSuggestIterator.class);
+    protected final static String SUGGEST_PREFIX = "suggest?term=";
+
+    private final ElasticIndexNode indexNode;
+    private final ElasticRequestHandler requestHandler;
+    private final ElasticResponseHandler responseHandler;
+    private final String suggestQuery;
+
+    private Iterator<FulltextResultRow> internalIterator;
+    private boolean loaded = false;
+
+    ElasticSuggestIterator(@NotNull ElasticIndexNode indexNode,
+                           @NotNull ElasticRequestHandler requestHandler,
+                           @NotNull ElasticResponseHandler responseHandler) {
+        this.indexNode = indexNode;
+        this.requestHandler = requestHandler;
+        this.responseHandler = responseHandler;
+        this.suggestQuery = requestHandler.getPropertyRestrictionQuery().replace(SUGGEST_PREFIX, "");
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (!loaded) {
+            try {
+                loadSuggestions();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+            loaded = true;
+        }
+        return internalIterator != null && internalIterator.hasNext();
+    }
+
+    @Override
+    public FulltextResultRow next() {
+        return internalIterator.next();
+    }
+
+    private void loadSuggestions() throws IOException {
+        BoolQueryBuilder suggestionQuery = requestHandler.suggestionMatchQuery(suggestQuery);
+        final SearchSourceBuilder searchSourceBuilder = SearchSourceBuilder.searchSource()
+                .query(suggestionQuery)
+                .size(100)
+                .fetchSource(FieldNames.PATH, null);
+        final SearchRequest searchRequest = new SearchRequest(indexNode.getDefinition().getRemoteIndexAlias())
+                .source(searchSourceBuilder);
+        SearchResponse res = indexNode.getConnection().getClient().search(searchRequest, RequestOptions.DEFAULT);
+        ArrayList<FulltextResultRow> results = new ArrayList<>();
+        PriorityQueue<ElasticSuggestion> pr = new PriorityQueue<>((a, b) -> Float.compare(b.score, a.score));
+        for (SearchHit doc : res.getHits()) {
+            if (responseHandler.isAccessible(responseHandler.getPath(doc))) {
+                for (SearchHit suggestion : doc.getInnerHits().get(":suggest").getHits()) {
+                    pr.add(new ElasticSuggestion(((List<String>) suggestion.getSourceAsMap().get("suggestion")).get(0), suggestion.getScore()));
+                }
+            }
+        }
+        for (ElasticSuggestion s : pr) {
+            results.add(new FulltextResultRow(s.suggestion));
+        }
+        this.internalIterator = results.iterator();
+    }
+
+    private final static class ElasticSuggestion {
+        String suggestion;
```

Comment:
I would make the fields "final" here... (I think using "final" for member variables does help, while I think it's not helpful for parameters and local variables). But it's a matter of taste.

Reply:
+1

Reply:
done